File: modules/backend/src/main/scala/cqrs/Backend.scala
/*
 * Copyright 2021 Hossein Naderi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package edomata.backend
package cqrs

import cats.effect.kernel.Async
import cats.effect.kernel.Resource
import edomata.core.*

import scala.concurrent.duration.*

trait Backend[F[_], S, R, N] {
  def compile: CommandHandler[F, S, N]
  def outbox: OutboxReader[F, N]
  def repository: RepositoryReader[F, S]
  def updates: NotificationsConsumer[F]
}

final class BackendBuilder[F[_]: Async, Codec[_], C, S, R, N] private[cqrs] (
    driver: Resource[F, StorageDriver[F, Codec]],
    domain: CQRSDomain[C, S, R, N],
    commandCache: Option[Resource[F, CommandStore[F]]],
    val maxRetry: Int = 5,
    val retryInitialDelay: FiniteDuration = 2.seconds
)(using StateModelTC[S]) {
  private def copy(
      driver: Resource[F, StorageDriver[F, Codec]] = driver,
      domain: CQRSDomain[C, S, R, N] = domain,
      commandCache: Option[Resource[F, CommandStore[F]]] = commandCache,
      maxRetry: Int = maxRetry,
      retryInitialDelay: FiniteDuration = retryInitialDelay
  ) = new BackendBuilder(
    driver = driver,
    domain = domain,
    maxRetry = maxRetry,
    commandCache = commandCache,
    retryInitialDelay = retryInitialDelay
  )

  def disableCache: BackendBuilder[F, Codec, C, S, R, N] =
    copy(commandCache = None)

  def withCommandCache(
      cache: Resource[F, CommandStore[F]]
  ): BackendBuilder[F, Codec, C, S, R, N] = copy(commandCache = Some(cache))

  def withCommandCache(
      cache: CommandStore[F]
  ): BackendBuilder[F, Codec, C, S, R, N] = withCommandCache(
    Resource.pure(cache)
  )

  def withCommandCache(
      cache: F[CommandStore[F]]
  ): BackendBuilder[F, Codec, C, S, R, N] = withCommandCache(
    Resource.eval(cache)
  )

  def withCommandCacheSize(
      maxCommandsToCache: Int
  ): BackendBuilder[F, Codec, C, S, R, N] = copy(commandCache =
    Some(Resource.eval(CommandStore.inMem(maxCommandsToCache)))
  )

  def withRetryConfig(
      maxRetry: Int = maxRetry,
      retryInitialDelay: FiniteDuration = retryInitialDelay
  ): BackendBuilder[F, Codec, C, S, R, N] =
    copy(maxRetry = maxRetry, retryInitialDelay = retryInitialDelay)

  def build(using
      state: Codec[S],
      notifs: Codec[N]
  ): Resource[F, Backend[F, S, R, N]] = for {
    d <- driver
    storage <- d.build[S, N, R]
  } yield new Backend[F, S, R, N] {

    override def compile: CommandHandler[F, S, N] =
      CommandHandler.withRetry(storage.repository, maxRetry, retryInitialDelay)

    override def outbox: OutboxReader[F, N] = storage.outbox

    override def repository: RepositoryReader[F, S] = storage.repository

    override def updates: NotificationsConsumer[F] = storage.updates

  }
}

final class PartialBackendBuilder[C, S, R, N] private[backend] (
    domain: CQRSDomain[C, S, R, N]
)(using model: StateModelTC[S]) {
  def use[F[_]: Async, Codec[_]](
      driver: StorageDriver[F, Codec]
  ): BackendBuilder[F, Codec, C, S, R, N] =
    from(Resource.pure(driver))

  def use[F[_]: Async, Codec[_], D <: StorageDriver[F, Codec]](
      driver: F[D]
  ): BackendBuilder[F, Codec, C, S, R, N] =
    from(Resource.eval(driver))

  def from[F[_]: Async, Codec[_], D <: StorageDriver[F, Codec]](
      driver: Resource[F, D]
  ): BackendBuilder[F, Codec, C, S, R, N] =
    new BackendBuilder(
      driver,
      domain,
      commandCache = Some(Resource.eval(CommandStore.inMem(1000)))
    )
}

File: rust/crates/edomata-backend/src/cqrs/backend.rs
//! The CQRS [`Backend`] and its builder.

use std::sync::Arc;

use edomata_core::{CommandMessage, CqrsDomainDsl, CqrsModel, Stomaton};

use super::{
    CachedRepository, CommandHandler, NotificationsConsumer, RepositoryReader, SharedStateModel,
    Storage, StorageDriver,
};
use crate::{
    BackendError, CommandStore, DomainService, InMemoryCommandStore, OutboxConsumer, OutboxItem,
    OutboxReader, Payload, RetryConfig,
};

/// A ready-to-use CQRS backend for one aggregate type.
///
/// ```
/// # use edomata_core::*;
/// # use edomata_backend::cqrs::Backend;
/// # use edomata_backend::inmemory::InMemoryDriver;
/// # struct Tally;
/// # impl CqrsModel for Tally { type State = i32; type Rejection = String; fn initial(&self) -> i32 { 0 } }
/// # tokio::runtime::Runtime::new().unwrap().block_on(async {
/// let dsl = Tally.dsl::<i32, String>();
/// let backend = Backend::builder(Tally, dsl)
///     .driver(InMemoryDriver::new())
///     .build_default()
///     .await
///     .unwrap();
///
/// let service = backend.compile(dsl.router(move |by| dsl.modify(move |s| s + by).void()));
/// let cmd = CommandMessage::new("cmd-1", chrono::Utc::now(), "tally-1", 5);
/// assert_eq!(service(cmd).await.unwrap(), Ok(()));
/// assert_eq!(backend.repository().get("tally-1").await.unwrap().state, 5);
/// # });
/// ```
pub struct Backend<S, N> {
    handler: CommandHandler<S, N>,
    outbox: Arc<dyn OutboxReader<N>>,
    repository: Arc<dyn RepositoryReader<S>>,
    updates: Arc<dyn NotificationsConsumer>,
}

impl<S, N> Clone for Backend<S, N> {
    fn clone(&self) -> Self {
        Self {
            handler: self.handler.clone(),
            outbox: Arc::clone(&self.outbox),
            repository: Arc::clone(&self.repository),
            updates: Arc::clone(&self.updates),
        }
    }
}

impl<S: Payload, N: Payload> Backend<S, N> {
    /// Starts building a backend for `model`. The `domain` marker fixes the
    /// notification type (its command and rejection types are irrelevant
    /// here).
    pub fn builder<M, C, R>(
        model: M,
        domain: CqrsDomainDsl<C, S, R, N>,
    ) -> PartialBackendBuilder<S, N>
    where
        M: CqrsModel<State = S> + Send + Sync + 'static,
    {
        let _ = domain;
        PartialBackendBuilder {
            model: Arc::new(model),
            _n: std::marker::PhantomData,
        }
    }

    /// Compiles a domain program into a service.
    pub fn compile<C: Payload, R: Payload>(
        &self,
        app: Stomaton<CommandMessage<C>, S, R, N, ()>,
    ) -> DomainService<C, R> {
        self.handler.compile(app)
    }

    /// The outbox.
    pub fn outbox(&self) -> &Arc<dyn OutboxReader<N>> {
        &self.outbox
    }

    /// The read side.
    pub fn repository(&self) -> &Arc<dyn RepositoryReader<S>> {
        &self.repository
    }

    /// Update signals.
    pub fn updates(&self) -> &Arc<dyn NotificationsConsumer> {
        &self.updates
    }

    /// Consumes the outbox with `handler`, now and whenever new items are
    /// published, until the backend is dropped. See [`OutboxConsumer`].
    pub async fn consume_outbox<F, Fut>(
        &self,
        consumer: OutboxConsumer,
        handler: F,
    ) -> Result<(), BackendError>
    where
        F: FnMut(OutboxItem<N>) -> Fut + Send,
        Fut: std::future::Future<Output = Result<(), BackendError>> + Send,
    {
        consumer
            .run(self.outbox.as_ref(), self.updates.outbox(), handler)
            .await
    }
}

/// First step of [`Backend::builder`]: choose a storage driver.
pub struct PartialBackendBuilder<S, N> {
    model: SharedStateModel<S>,
    _n: std::marker::PhantomData<fn() -> N>,
}

impl<S: Payload, N: Payload> PartialBackendBuilder<S, N> {
    /// Uses `driver` to store this aggregate.
    pub fn driver<D: StorageDriver>(self, driver: D) -> BackendBuilder<S, N, D> {
        BackendBuilder {
            model: self.model,
            driver,
            command_cache: CommandCacheChoice::InMemory(1000),
            handler: None,
            retry: RetryConfig::default(),
            state_cache_size: 1000,
        }
    }
}

enum CommandCacheChoice {
    Disabled,
    InMemory(usize),
    Custom(Arc<dyn CommandStore>),
}

/// Configures and builds a CQRS [`Backend`]. Mirrors Scala's
/// `cqrs.BackendBuilder`.
///
/// Defaults: an in-memory command cache of 1000 commands, a state cache of
/// 1000 aggregates, and [`RetryConfig::default`].
pub struct BackendBuilder<S, N: 'static, D: StorageDriver> {
    model: SharedStateModel<S>,
    driver: D,
    command_cache: CommandCacheChoice,
    handler: Option<D::Handler<N>>,
    retry: RetryConfig,
    state_cache_size: usize,
}

impl<S: Payload, N: Payload, D: StorageDriver> BackendBuilder<S, N, D> {
    /// Disables command caching (every command is loaded from the storage).
    /// This also disables the state cache.
    pub fn disable_cache(mut self) -> Self {
        self.command_cache = CommandCacheChoice::Disabled;
        self
    }

    /// Uses a custom command store.
    pub fn with_command_cache(mut self, store: Arc<dyn CommandStore>) -> Self {
        self.command_cache = CommandCacheChoice::Custom(store);
        self
    }

    /// Caches the ids of the last `max_commands_to_cache` commands in
    /// memory.
    pub fn with_command_cache_size(mut self, max_commands_to_cache: usize) -> Self {
        self.command_cache = CommandCacheChoice::InMemory(max_commands_to_cache);
        self
    }

    /// Number of aggregate states cached in memory.
    pub fn with_state_cache_size(mut self, size: usize) -> Self {
        self.state_cache_size = size;
        self
    }

    /// Sets the retry policy for version conflicts.
    pub fn with_retry_config(mut self, retry: RetryConfig) -> Self {
        self.retry = retry;
        self
    }

    /// Runs `handler` inside the save transaction for every batch of
    /// notifications (for example to maintain projections).
    pub fn with_event_handler(mut self, handler: D::Handler<N>) -> Self {
        self.handler = Some(handler);
        self
    }

    /// The configured retry policy.
    pub fn retry_config(&self) -> RetryConfig {
        self.retry
    }

    /// Builds the backend with default codecs.
    pub async fn build_default(self) -> Result<Backend<S, N>, BackendError>
    where
        D::Codec<S>: Default,
        D::Codec<N>: Default,
    {
        self.build(Default::default(), Default::default()).await
    }

    /// Builds the backend.
    pub async fn build(
        self,
        state_codec: D::Codec<S>,
        notification_codec: D::Codec<N>,
    ) -> Result<Backend<S, N>, BackendError> {
        let Storage {
            repository,
            outbox,
            updates,
        } = self
            .driver
            .build(
                Arc::clone(&self.model),
                state_codec,
                notification_codec,
                self.handler,
            )
            .await?;
        let repository: Arc<dyn super::Repository<S, N>> = match self.command_cache {
            CommandCacheChoice::Disabled => repository,
            CommandCacheChoice::InMemory(size) => Arc::new(CachedRepository::with_size(
                repository,
                Arc::new(InMemoryCommandStore::new(size)),
                self.state_cache_size,
            )),
            CommandCacheChoice::Custom(store) => Arc::new(CachedRepository::with_size(
                repository,
                store,
                self.state_cache_size,
            )),
        };
        let reader: Arc<dyn RepositoryReader<S>> = Arc::new(ReaderOf(Arc::clone(&repository)));
        Ok(Backend {
            handler: CommandHandler::with_retry(repository, self.retry),
            outbox,
            repository: reader,
            updates,
        })
    }
}

/// Exposes the read side of a repository.
struct ReaderOf<S, N>(Arc<dyn super::Repository<S, N>>);

#[async_trait::async_trait]
impl<S: Payload, N: Payload> RepositoryReader<S> for ReaderOf<S, N> {
    async fn get(&self, id: &str) -> Result<super::AggregateState<S>, BackendError> {
        self.0.get(id).await
    }
}
